addon/openssl-crypto.cc

```diff
@@ -0,0 +1,457 @@
+// Adapted from
+// https://github.com/mongodb/libmongocrypt/blob/18cb9e4e900c45c0b9b71fd34e159f8cb29fe1de/src/crypto/libcrypto.c
+// and
+// https://github.com/mongodb/libmongocrypt/blob/18cb9e4e900c45c0b9b71fd34e159f8cb29fe1de/kms-message/src/kms_crypto_libcrypto.c
+// This file provides native crypto hooks for OpenSSL 3 (the default since Node.js 18),
+// allowing us to skip expensive round-trips between JS and C++.
+
+#include "mongocrypt.h"
+
+// Electron does not expose OpenSSL, so we cannot use OpenSSL
+// functions directly if we're building against Electron:
+// https://github.com/electron/electron/issues/13176
+#ifndef MONGOCRYPT_AVOID_OPENSSL_CRYPTO
+#include <openssl/crypto.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+#include <openssl/rand.h>
+
+#include <stdexcept>
+
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <dlfcn.h>
+#endif
+
+#undef ASSERT
+#undef STRINGIFY
+#define STRINGIFY(x) #x
+#define ASSERT(x)                                                             \
+    do {                                                                      \
+        if (!(x)) {                                                           \
+            throw std::runtime_error("Assertion failed: " #x " (at " __FILE__ \
+                                     ":" STRINGIFY(__LINE__) ")");            \
+        }                                                                     \
+    } while (0)
+
+// Helpers to easily access OpenSSL symbols in a type-safe way.
+#ifdef MONGO_CLIENT_ENCRYPTION_STATIC_OPENSSL
+#define S_Unchecked(x) (x)
+#define S(x) S_Unchecked(x)
+#else
+#define S_Unchecked(x)                                                           \
+    ([&]() {                                                                     \
+        static void* const sym = node_mongocrypt::opensslcrypto::opensslsym(#x); \
+        return reinterpret_cast<decltype(x)*>(sym);                              \
+    })()
+#define S(x)                                                                       \
```

Comment:
No particularly strong feelings on the naming, although this symbol does get used quite a bit and _should_ seem like it isn't very relevant while scanning the code (e.g. `S(EVP_DigestInit_ex)(digest_ctxp, hash, nullptr)` looking similar to `EVP_DigestInit_ex(digest_ctxp, hash, nullptr)` is a good thing imo, because those do the same things, it's just how we access the function that differs here)

> ```c++
> auto encrypt_context_init = GET_OPENSSL_SYMBOL(EVP_EncryptInit_ex);
> ```

If we do introduce this pattern here, let's keep the identifiers associated with each other, e.g. `EVP_EncryptInit_ex_fn` instead of `encrypt_context_init`, otherwise code searches for `EVP_EncryptInit_ex` won't show its actual usage anymore 🙂

Reply:
Hm, I thought I responded to this but I guess my comment got lost.

I was thinking that the intention with the macro was, in addition to reducing duplication, to abstract raw usage of openssl, but it appears that wasn't the intention.

I don't feel strongly so I'm going to just resolve this comment.

Reply:
Yeah, not the intention, I personally quite like that this is not too different from the original libmongocrypt code so that it's easier to compare the two